src/types.rs

```diff
@@ -522,51 +534,135 @@ pub enum DateError {
     /// The month is out of bounds.
     #[error("month not in interval 1-12")]
     MonthOutOfBounds,
+    /// The offset is out of bounds.
+    #[error("offset not in interval -24:00 -- 24:00")]
+    OffsetOutOfBounds,
+}
+
+impl Display for Date {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let Date {
+            year,
+            month,
+            day,
+            hour,
+            minute,
+            second,
+            nanosecond,
+            offset,
+        } = self;
+        write!(f, "{year:04}")?;
+
+        if let Some(month) = month {
+            let month = month + 1;
+            write!(f, "-{month:02}")?;
+        } else {
+            return Ok(());
+        }
+
+        if let Some(day) = day {
+            let day = day + 1;
+            write!(f, "-{day:02}")?;
+        } else {
+            return Ok(());
+        }
+
+        if let Some(hour) = hour {
+            write!(f, "T{hour:02}")?;
+        } else {
+            return Ok(());
+        }
+
+        if let Some(minute) = minute {
+            write!(f, ":{minute:02}")?;
+            if let Some(second) = second {
+                write!(f, ":{second:02}")?;
+                if let Some(nanosecond) = nanosecond {
+                    write!(f, ".{nanosecond:09}")?;
+                }
+            }
+        }
```

Comment:
We could probably replace it with a match on all three variables, but that would probably involve some duplication. Or we could move to an enum layout for `Date` then this would also turn into a match.

Reply:
You could have guard clauses for an early return on each but that's about twice as much code and quite ugly.